sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
+        logits_d, m_d, prec_d, _, _ = dist.get_mixture_components(encoded_x)
+        norm_logits_d = logits_d - torch.logsumexp(logits_d, dim=-1, keepdim=True)
+
+        # z-score theta if it z-scoring had been requested.
+        theta = self._maybe_z_score_theta(theta)
+
+        # Compute the MoG parameters of the proposal posterior.
+        logits_pp, m_pp, prec_pp, cov_pp = self._automatic_posterior_transformation(
+            norm_logits_p, m_p, prec_p, norm_logits_d, m_d, prec_d,
+        )
+
+        # Compute the log_prob of theta under the product.
+        log_prob_proposal_posterior = _mog_log_prob(
+            theta, logits_pp, m_pp, cov_pp, prec_pp,
+        )
+        # print(logits_pp)
+        self._assert_all_finite(log_prob_proposal_posterior, "proposal posterior eval")
+
+        return log_prob_proposal_posterior
+
+    def _automatic_posterior_transformation(
+        self,
+        logits_p: Tensor,
+        means_p: Tensor,
+        precisions_p: Tensor,
+        logits_d: Tensor,
+        means_d: Tensor,
+        precisions_d: Tensor,
+    ):
+        r"""
+        Returns the MoG parameters of the proposal posterior.
+
+        The proposal posterior is:
+        $pp(\theta|x) = 1/Z * q(\theta|x) * prop(\theta) / p(\theta)$
+        In words: proposal posterior = posterior estimate * proposal / prior.
+
+        If the posterior estimate and the proposal are MoG and the prior is either
+        Gaussian or uniform, we can solve this in closed-form. The is implemented in
+        this function.
+
+        We have to build L*K components. How do we do this?
+        Example: proposal has two components, density estimator has three components.
+        Let's call the two components of the proposal i,j and the three components
+        of the density estimator x,y,z. We have to multiply any component of the
+        proposal with any component of the density estimator. So, what we do is:
+        1) for the proposal, build: i,i,i,j,j,j. Done with torch.repeat_interleave()
+        2) for the density estimator, build: x,y,z,x,y,z. Done with torch.repeat()
+        3) Multiply them with simple matrix operations.
+
+        Args:
+            logits_p: Component weight of each Gaussian of the proposal.
+            means_p: Mean of each Gaussian of the proposal.
+            precisions_p: Precision matrix of each Gaussian of the proposal.
+            logits_d: Component weight for each Gaussian of the density estimator.
+            means_d: Mean of each Gaussian of the density estimator.
+            precisions_d: Precision matrix of each Gaussian of the density estimator.
+
+        Returns: (Component weight, mean, precision matrix, covariance matrix) of each
+            Gaussian of the proposal posterior. Has L*K terms (proposal has L terms,
+            density estimator has K terms).
+        """
+
+        precisions_pp, covariances_pp = self._precisions_proposal_posterior(
```

Comment:
this modularity is great!

Reply:
Maybe add a comment at the beginning that we are collecting the terms we need to apply the formulas from the APT paper appendix A1 or what